### PR TITLE
Teleportation API

### PR DESCRIPTION
**Pull request by Owen1212055**
This expands the teleport API to allow for teleporting entities while keeping passengers and allowing to teleport the player while keeping the player mounted.

:star: Teleporting player while keeping passengers and dismounting
https://user-images.githubusercontent.com/23108066/132114747-25568ff0-7f46-4186-84ce-064903781465.mp4

The idea of this PR is to essentially allow
1. Keeping passengers when teleporting entities

2. Allowing the entity being teleported to stay mounted when teleported
- This causes the player to teleport for a moment, but because they are still mounted it snaps them back. However because of the new packet we can now choose if we should keep the player mounted or dismount them. I am not sure if we should return false here OR allow this. See: https://bugs.mojang.com/browse/MC-202202

Behavior (keepPassengers, dismount):
- Teleporting entity  (false, true) DEFAULT BEHAVIOR: If the entity has passengers it will return false (legacy behavior), else it will kick them off their vehicle.
- Teleporting entity (false, false): If the entity has passengers it will return false, but it will not kick them off the vehicle
- Teleporting entity (true, false): This will teleport the entity and most likely cause it to be teleported back next tick because it has a vehicle attached. (Passengers and vehicle will not be affected)
- Teleporting entity (true, true) This will teleport the entity and all passengers ontop with them. The entity teleported will be dismounted

So basically it will return false if keepPassengers is false and they have passengers

Tested with players and entities, seems to work just fine. Feedback appreciated!

### CODE REVIEW

**Comment by Owen1212055** [comment]
Added new API for support relative teleportation + look at api.
This has special behavior on players, where you can prevent resetting the players velocity while still forcing the player to be moved.

(Code for handling the position packet flags on client)
```java
 boolean hasXFlag = packet.getFlags().contains(PlayerPositionLookS2CPacket.Flag.X);
      boolean hasYFlag = packet.getFlags().contains(PlayerPositionLookS2CPacket.Flag.Y);
      boolean hasZFlag = packet.getFlags().contains(PlayerPositionLookS2CPacket.Flag.Z);
      double newXVelocity;
      double newXCoord;
      if (hasXFlag) {
         newXVelocity = vec3d.getX();
         newXCoord = playerEntity.getX() + packet.getX();
         playerEntity.lastRenderX += packet.getX();
      } else {
         newXVelocity = 0.0D;
         newXCoord = packet.getX();
         playerEntity.lastRenderX = newXCoord;
      }

      double newYVelocity;
      double newYCoord;
      if (hasYFlag) {
         newYVelocity = vec3d.getY();
         newYCoord = playerEntity.getY() + packet.getY();
         playerEntity.lastRenderY += packet.getY();
      } else {
         newYVelocity = 0.0D;
         newYCoord = packet.getY();
         playerEntity.lastRenderY = newYCoord;
      }

      double newZVelocity;
      double newZCoord;
      if (hasZFlag) {
         newZVelocity = vec3d.getZ();
         newZCoord = playerEntity.getZ() + packet.getZ();
         playerEntity.lastRenderZ += packet.getZ();
      } else {
         newZVelocity = 0.0D;
         newZCoord = packet.getZ();
         playerEntity.lastRenderZ = newZCoord;
      }
```

This allows you to change a specific coordinate without causing the entire player's velocity to be reset. For example, setting the player's yaw/pitch while allowing the player to move freely.
https://user-images.githubusercontent.com/23108066/132130745-f8d22752-c8b2-42ed-9f47-96f492901f52.mp4


Unsure if this should be in the same patch, but this also includes **look at** api for players.
This exposes a new lookat packet (used in /teleport command), which causes the player to look at a location or entity.
https://user-images.githubusercontent.com/23108066/132131985-5a92ef5e-4bb2-4259-b779-37fad56bafae.mp4

**Review by lynxplay — COMMENTED** [review]
If the concept of relative teleportation flags is introduced, should those also be mirrored (be modified by) the [PlayerTeleportEvent](https://papermc.io/javadocs/paper/1.17/org/bukkit/event/player/PlayerTeleportEvent.html) ?

**Comment by Owen1212055** [comment]
> If the concept of relative teleportation flags is introduced, should those also be mirrored (be modified by) the [PlayerTeleportEvent](https://papermc.io/javadocs/paper/1.17/org/bukkit/event/player/PlayerTeleportEvent.html) ?

Looking into this a bit more, this would require a bit more changes than I would like here. As alot of methods just hide these parameters (see this one)
image

Additionally, it seems in some places the relative flags are removed completely. So, opinions would be appreciated.

**Comment by lynxplay** [comment]
> So, opinions would be appreciated.

I agree, I don't think they are a 100% necessity especially with how incredibly niche the flags are anyway.

At least the usage of these should be documented better in terms of javadocs though. Reading `"Coordinates of the location that should be relative instead of absolute"` does not, at least to me, convey the fact that the following:
```java
var finalLocation = new Location(player.getWorld(), 0, 1, 0, 0, 0);
player.teleport(finalLocation, PlayerTeleportEvent.TeleportCause.PLUGIN, false, false, RelativeTeleportFlag.values());
```
does still teleport the player to [0,1,0] and only has a client side effect in regards to cancelled motion. While event additions might not be required due to how few teleportations will actually use these (and want others to mutate them), documentation then becomes just the more important.

**Comment by Owen1212055** [comment]
> > So, opinions would be appreciated.
> 
> I agree, I don't think they are a 100% necessity especially with how incredibly niche the flags are anyway.
> 
> At least the usage of these should be documented better in terms of javadocs though. Reading `"Coordinates of the location that should be relative instead of absolute"` does not, at least to me, convey the fact that the following:
> 
> ```java
> var finalLocation = new Location(player.getWorld(), 0, 1, 0, 0, 0);
> player.teleport(finalLocation, PlayerTeleportEvent.TeleportCause.PLUGIN, false, false, RelativeTeleportFlag.values());
> ```
> 
> does still teleport the player to [0,1,0] and only has a client side effect in regards to cancelled motion. While event additions might not be required due to how few teleportations will actually use these (and want others to mutate them), documentation then becomes just the more important.

Ahhh, so perhaps I did not actually describe it well enough in my PR. This does cause the player to be moved relatively 0 1 0. The behavior can be seen by using the teleport command and using /teleport @p ~ ~1 ~ 

Perhaps then a new type of teleport like ``RelativeTeleportEvent`` could be added which includes these flags? This way we could prevent breaking any API by simply making it so the getLocation will be made absolute and then we can add a method to get the actual relative location and flags.

**Comment by lynxplay** [comment]
> Ahhh, so perhaps I did not actually describe it well enough in my PR. This does cause the player to be moved relatively 0 1 0. The behavior can be seen by using the teleport command and using /teleport @p ~ ~1 ~

Did you actually test this behaviour then ? Looking at the `internalTeleport` it seems to expect the final location not the relative. The modifications based on the flags are only applied to the packet send, the actual location is based directly on the location passed in `d0, d1,d2`:
```java
this.awaitingTeleportTime = this.tickCount;
this.player.moveTo(d0, d1, d2, f, f1); // Paper - use proper setPositionRotation for teleportation
this.player.forceCheckHighPriority(); // Paper
this.player.connection.send(new ClientboundPlayerPositionPacket(d0 - d3, d1 - d4, d2 - d5, f - f2, f1 - f3, set, this.awaitingTeleport, flag));
```

Your method still just passed through the then "relative" or partially "relative" location, resulting in the behaviour I described above.
Could you please validate that this isn't just something I somehow screwed up when checking out your PR?

Example listener showing the issue, teleporting the player to [0,0,0] instead of keeping them at their current position.
```java
@EventHandler
public void on(final PlayerDropItemEvent event) {
    var relativeMotion = new Location(event.getPlayer().getWorld(), 0, 0, 0);
    event.getPlayer().teleport(relativeMotion, PlayerTeleportEvent.TeleportCause.PLUGIN, true, false, RelativeTeleportFlag.values());
}
```

**Comment by Owen1212055** [comment]
> > Ahhh, so perhaps I did not actually describe it well enough in my PR. This does cause the player to be moved relatively 0 1 0. The behavior can be seen by using the teleport command and using /teleport @p ~ ~1 ~
> 
> Did you actually test this behaviour then ? Looking at the `internalTeleport` it seems to expect the final location not the relative. The modifications based on the flags are only applied to the packet send, the actual location is based directly on the location passed in `d0, d1,d2`:
> 
> ```java
> this.awaitingTeleportTime = this.tickCount;
> this.player.moveTo(d0, d1, d2, f, f1); // Paper - use proper setPositionRotation for teleportation
> this.player.forceCheckHighPriority(); // Paper
> this.player.connection.send(new ClientboundPlayerPositionPacket(d0 - d3, d1 - d4, d2 - d5, f - f2, f1 - f3, set, this.awaitingTeleport, flag));
> ```
> 
> Your method still just passed through the then "relative" or partially "relative" location, resulting in the behaviour I described above.
> Could you please validate that this isn't just something I somehow screwed up when checking out your PR?
> 
> Example listener showing the issue, teleporting the player to [0,0,0] instead of keeping them at their current position.
> 
> ```java
> @EventHandler
> public void on(final PlayerDropItemEvent event) {
>     var relativeMotion = new Location(event.getPlayer().getWorld(), 0, 0, 0);
>     event.getPlayer().teleport(relativeMotion, PlayerTeleportEvent.TeleportCause.PLUGIN, true, false, RelativeTeleportFlag.values());
> }
> ```

You are totally correct here. I think I did some really botched testing, so I appreciate you double-checking me here. 
Let me improve this documentation to make this a bit more clear.

Again, thank you very much.

**Comment by Owen1212055** [comment]
I tried changing the documentation, do you have any feedback?

**Comment by lynxplay** [comment]
> I tried changing the documentation, do you have any feedback?

I like it :+1: could you reflect the additional information in the parameter description ?
E.g.:
```patch
-+ * @param teleportFlags Coordinates of the location that should be relative instead of absolute
++ * @param teleportFlags Coordinates of the location that the client should handle as relative teleportation.
``` 
Just to ensure no one actually reads it and believes the flags cause any form of mutation to the final location. Your addition already covered it but half of the people barely read the javadocs, let alone something after the first `<p>` tag :sweat_smile:

**Comment by Owen1212055** [comment]
Fair, thank you for all ur feedback lynx. I do appreciate it.

File: patches/server/0804-More-Teleport-API.patch [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
++
++    @Override
++    public boolean teleport(@org.jetbrains.annotations.NotNull Location location, boolean keepPassengers, boolean dismount) {
++        return teleport(location, TeleportCause.PLUGIN, keepPassengers, dismount);
++    }
++
++    @Override
++    public boolean teleport(@org.jetbrains.annotations.NotNull Location location, @org.jetbrains.annotations.NotNull TeleportCause cause, boolean keepPassengers) {
++        return teleport(location, keepPassengers, true);
++    }
```
Comment (kennytv) [inline patches/server/0804-More-Teleport-API.patch]: Use default interface methods

File: patches/server/0804-More-Teleport-API.patch [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+     // Paper end
+ 
+     @Override
+-    public boolean teleport(Location location, PlayerTeleportEvent.TeleportCause cause) {
++    public void lookAt(@NotNull io.papermc.paper.entity.LookAnchor playerAnchor, org.bukkit.entity.@NotNull Entity entity, @NotNull io.papermc.paper.entity.LookAnchor entityAnchor) {
++        net.minecraft.commands.arguments.EntityAnchorArgument.Anchor[] anchors = net.minecraft.commands.arguments.EntityAnchorArgument.Anchor.values();
++        getHandle().lookAt(anchors[playerAnchor.ordinal()], ((CraftEntity) entity).getHandle(), anchors[entityAnchor.ordinal()]);
++    }
++
++    @Override
```
Comment (kennytv) [inline patches/server/0804-More-Teleport-API.patch]: Default method

File: patches/server/0804-More-Teleport-API.patch [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
++    public boolean teleport(@org.jetbrains.annotations.NotNull Location location, boolean keepPassengers, boolean dismount) {
++        return teleport(location, TeleportCause.PLUGIN, keepPassengers, dismount);
++    }
++
++    @Override
++    public boolean teleport(@org.jetbrains.annotations.NotNull Location location, @org.jetbrains.annotations.NotNull TeleportCause cause, boolean keepPassengers) {
++        return teleport(location, keepPassengers, true);
```
Comment (kennytv) [inline patches/server/0804-More-Teleport-API.patch]: What happened to the cause here?

File: patches/api/0332-More-Teleport-API.patch [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
++     * Teleports this entity to the given location.
++     * <p>
++     * Note: You cannot keep passengers or stay mounted if you are teleporting this entity to another world.
++     *
++     * <p>
++     * Relative teleportation flags are only used client side, and cause the player to not lose velocity in that
++     * specific coordinate. The location of the teleportation will not change.
++     *
++     * @param location New location to teleport this entity to
++     * @param keepPassengers If passengers should be kept when teleporting
```
Comment (kennytv) [inline patches/api/0332-More-Teleport-API.patch]: The description here is incorrect; it's not about whether they should be kept, but whether teleporting should be possible *at all* with passengers (same for everywhere else)

File: patches/api/0332-More-Teleport-API.patch [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
++    /**
++     * Teleports this entity to the given location.
++     * <p>
++     * Note: You cannot keep passengers or stay mounted if you are teleporting this entity to another world.
++     *
++     * <p>
++     * Relative teleportation flags are only used client side, and cause the player to not lose velocity in that
++     * specific coordinate. The location of the teleportation will not change.
++     *
++     * @param location New location to teleport this entity to
```
Comment (kennytv) [inline patches/api/0332-More-Teleport-API.patch:137]: New or relative

File: patches/api/0332-More-Teleport-API.patch [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
++++ b/src/main/java/org/bukkit/entity/Player.java
+@@ -2289,6 +2289,52 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
+     String getClientBrandName();
+     // Paper end
+ 
++    // Paper start - Teleport API
++    /**
++     * Teleports this entity to the given location.
++     * <p>
++     * Note: You cannot keep passengers or stay mounted if you are teleporting this entity to another world.
```
Comment (kennytv) [inline patches/api/0332-More-Teleport-API.patch]: Move this into an `@throws` tag (same for everywhere else)

**Review by kennytv — CHANGES_REQUESTED** [review]

File: patches/server/0804-More-Teleport-API.patch (hunk 1 above)
Comment (Owen1212055) [reply patches/server/0804-More-Teleport-API.patch]: I was looking before and generally I never saw default methods in interfaces, only a couple places which some had a // TODO: Move to implementation 

Is there a reason why they aren’t used often? I can still do it but I’m curious.

Comment (kennytv) [reply patches/server/0804-More-Teleport-API.patch]: Probably because Spigot is scared of using anything new, including Java features, and because people tried keeping things similar to it - which is not a goal we should follow if it goes against simplicity/readability. Default methods definitely are used a good bunch in Paper API as well

**Comment by Owen1212055** [comment]
Rebased and resolved. Changed keep passengers to ignore passengers, added throw javadoc and added stuff as default interface methods.

**Comment by Owen1212055** [comment]
Shoo

**Comment by Owen1212055** [comment]
rebased

**Comment by amaranth** [comment]
Is there a reason you can't teleport with a vehicle or passenger to another world? If there is no technical reason stopping this the extra work might be worth considering to make the API simpler to use. If that can't be done I think the API should at least just return false in this case instead of throwing an exception.

**Comment by Owen1212055** [comment]
> Is there a reason you can't teleport with a vehicle or passenger to another world? If there is no technical reason stopping this the extra work might be worth considering to make the API simpler to use. If that can't be done I think the API should at least just return false in this case instead of throwing an exception.

I’ll have to play around, I didn’t try messing with that too much as generally looking at how other people did cross world teleportation it seemed to be quite a large task. So, ill see if it works nicely or not.

**Comment by Owen1212055** [comment]
> Is there a reason you can't teleport with a vehicle or passenger to another world? If there is no technical reason stopping this the extra work might be worth considering to make the API simpler to use. If that can't be done I think the API should at least just return false in this case instead of throwing an exception.

The Minecraft method itself forcibly removes passengers.. and judging how there is no logic for handling them either I don't think doing this is very safe.

**Comment by Owen1212055** [comment]
Rebased

**Comment by Owen1212055** [comment]
Rebased

**Comment by Owen1212055** [comment]
Rebased and cleaned up the diff

File: patches/api/0374-More-Teleport-API.patch [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
++    }
++
++    /**
++     * Returns the relative teleportation flags used in this teleportation.
++     * This determines which axis the player will not lose their velocity in.
++     *
++     * @return relative teleportation flags
++     */
++    @NotNull
++    public java.util.Set<io.papermc.paper.entity.RelativeTeleportFlag> getRelativeTeleportationFlags() {
```
Comment (lynxplay) [inline patches/api/0374-More-Teleport-API.patch]: Either we make this set immutable or we clearly state that mutation of the set will not affect teleportation.
If I just had a first glance at the event as a developer and I see a set returned like this (e.g. like blockList in explosion events) I'd assumable mutability and said mutations to affect the outcome.

Technically this could be mutable ? Might just be a point to discuss :+1:

**Comment by Astralchroma** [comment]
You asked for everyone *(including #paper-contrib stalkers like me)* to give feedback so here we go:

Looks good, I don't see anything that immediately stands out as being an issue.

This would be great for me, as I currently acomplish this kind of teleportation through NMS and reflection which is a real pain to deal with.

**Comment by maestro-denery** [comment]
Yeah, I think it would be nice addition, it useful when making "vanilla-like" teleport things, for example I have a plugin adding "WayStones", which teleport players to another connected "waystone" if they interact with it. This API will help in this case. It is my feedback, thanks for making such things!

File: patches/api/0374-More-Teleport-API.patch (hunk 7 above)
Comment (Owen1212055) [reply patches/api/0374-More-Teleport-API.patch]: Yea, I guess making it mutable is fine.

Comment (lynxplay) [reply patches/api/0374-More-Teleport-API.patch]: I mean, idk how useful mutability is here. A listener mutating this seems pretty unlikely. Just something to choose here.

//edit[0]: also idk if actual mutability is desirable, rather setter and getter if you want to (tho again, really don't know if you want to in the first place)

File: patches/api/0374-More-Teleport-API.patch [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+new file mode 100644
+index 0000000000000000000000000000000000000000..d29f7b10f86d049e935d614c67c9c2e57287d77e
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/entity/LookAnchor.java
+@@ -0,0 +1,16 @@
++package io.papermc.paper.entity;
++
++import org.bukkit.entity.LivingEntity;
++
++public enum LookAnchor {
```
Comment (Machine-Maker) [inline patches/api/0374-More-Teleport-API.patch]: add javadoc comment to the type. like what this is used for.

File: patches/api/0374-More-Teleport-API.patch [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
++}
+diff --git a/src/main/java/io/papermc/paper/entity/RelativeTeleportFlag.java b/src/main/java/io/papermc/paper/entity/RelativeTeleportFlag.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..4270deba5f4cd85a6789fcd190110f9b06ebf310
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/entity/RelativeTeleportFlag.java
+@@ -0,0 +1,25 @@
++package io.papermc.paper.entity;
++
++public enum RelativeTeleportFlag {
```
Comment (Machine-Maker) [inline patches/api/0374-More-Teleport-API.patch]: Add javadoc comment to type for what this is for.

File: patches/server/0878-More-Teleport-API.patch [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+-        PlayerTeleportEvent event = new PlayerTeleportEvent(player, from.clone(), to.clone(), cause);
++        // Paper start - Teleport API
++        Set<io.papermc.paper.entity.RelativeTeleportFlag> relativeFlags = new java.util.HashSet<>(set.size());
++        for (net.minecraft.network.protocol.game.ClientboundPlayerPositionPacket.RelativeArgument relativeArgument : set) {
++            relativeFlags.add(switch (relativeArgument) {
++                case X -> io.papermc.paper.entity.RelativeTeleportFlag.X;
++                case Y -> io.papermc.paper.entity.RelativeTeleportFlag.Y;
++                case Z -> io.papermc.paper.entity.RelativeTeleportFlag.Z;
++                case X_ROT -> io.papermc.paper.entity.RelativeTeleportFlag.PITCH;
++                case Y_ROT -> io.papermc.paper.entity.RelativeTeleportFlag.YAW;
```
Comment (Machine-Maker) [inline patches/server/0878-More-Teleport-API.patch]: I'd throw this in a static util method somewhere. on the chance this conversion might need to happen elsewhere in the future.

**Comment by Owen1212055** [comment]
Updated with suggestions: https://canary.discord.com/channels/289587909051416579/925530366192779286/953810980863893504

**Review by Machine-Maker — DISMISSED** [review]

**Comment by kennytv** [comment]
I don't see the point of having a dismount boolean is setting it to false results in an instant teleport back anyways? That special case just seems to be confusing, and if people want the bavior of false, false, they should just check if it has passengers themselves

File: patches/server/0878-More-Teleport-API.patch [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
++    }
++
++    public static io.papermc.paper.entity.RelativeTeleportFlag toApiRelativeFlag(net.minecraft.network.protocol.game.ClientboundPlayerPositionPacket.RelativeArgument nmsFlag) {
++        return switch (nmsFlag) {
++            case X -> io.papermc.paper.entity.RelativeTeleportFlag.X;
++            case Y -> io.papermc.paper.entity.RelativeTeleportFlag.Y;
++            case Z -> io.papermc.paper.entity.RelativeTeleportFlag.Z;
++            case X_ROT -> io.papermc.paper.entity.RelativeTeleportFlag.PITCH;
++            case Y_ROT -> io.papermc.paper.entity.RelativeTeleportFlag.YAW;
++        }
```
Comment (kennytv) [inline patches/server/0878-More-Teleport-API.patch]:
```suggestion
+        };
```

**Review by kennytv — DISMISSED** [review]

**Comment by Owen1212055** [comment]
> I don't see the point of having a dismount boolean is setting it to false results in an instant teleport back anyways? That special case just seems to be confusing, and if people want the bavior of false, false, they should just check if it has passengers themselves

The reason for this is special behavior, it basically allows you to teleport and modify the place even while they are still riding the entity. Yes this is niche, however, it is still more or less functionality that was implemented by Mojang after servers were unable to do this behavior anymore due to entity riding being moved more client side.

**Comment by Owen1212055** [comment]
Opinions appreciated.
~~Going to move the dismounted boolean to its own method, as currently the default behavior is to dismount anyways.
**teleportMounted**~~

- Unsure about this one now, as it seems to have useful behavior to reduce stuttering when teleporting entities.  See https://canary.discord.com/channels/289587909051416579/555462289851940864/955149016692322374

Additionally, by using the same behavior as faceTo to nicely add a method to set the players rotation without messing with velocity. Any feedback on implementations would be appreciated as I'd most likely be shifting a location at a very far distance in the direction (faceTo only uses a block coordinate).

File: patches/server/0878-More-Teleport-API.patch [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
++    }
++
++    @Override
++    public boolean teleport(Location location, PlayerTeleportEvent.TeleportCause cause, boolean ignorePassengers, boolean dismount) {
++        return teleport(location, cause, ignorePassengers, dismount, new io.papermc.paper.entity.RelativeTeleportFlag[0]);
++    }
++
++    @Override
++    public void lookAt(@NotNull org.bukkit.entity.Entity entity, @NotNull io.papermc.paper.entity.LookAnchor playerAnchor, @NotNull io.papermc.paper.entity.LookAnchor entityAnchor) {
++        net.minecraft.commands.arguments.EntityAnchorArgument.Anchor[] anchors = net.minecraft.commands.arguments.EntityAnchorArgument.Anchor.values();
```
Comment (montlikadani) [inline patches/server/0878-More-Teleport-API.patch]: You could cache this array statically to avoid copying the `Anchor` values each time this method is called.

File: patches/server/0878-More-Teleport-API.patch [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
++        getHandle().lookAt(anchors[playerAnchor.ordinal()], ((CraftEntity) entity).getHandle(), anchors[entityAnchor.ordinal()]);
++    }
++
++    @Override
++    public void lookAt(double x, double y, double z, @NotNull io.papermc.paper.entity.LookAnchor playerAnchor) {
++        this.getHandle().lookAt(net.minecraft.commands.arguments.EntityAnchorArgument.Anchor.values()[playerAnchor.ordinal()], new Vec3(x, y, z));
```
Comment (montlikadani) [inline patches/server/0878-More-Teleport-API.patch]: Same as above

File: patches/server/0878-More-Teleport-API.patch (hunk 12 above)
Comment (lynxplay) [reply patches/server/0878-More-Teleport-API.patch]: Could also just add a simple switch statement converted, similar to `toApiRelativeFlag` or `toNmsRelativeFlag`, which would a) avoid the 2-pointer array allocation caused by calling #values() and b) removes the dependency on the enums to be ordered alike.

Comment (Owen1212055) [reply patches/server/0878-More-Teleport-API.patch]: Done

**Comment by Owen1212055** [comment]
Implemented Entity#setRotation for players, I am not a mathy guy... but opinions on the implementation here would be appreciated.

**Comment by Owen1212055** [comment]
Rebased and fixed funky multiply in set rotation for players

File: patches/api/0382-More-Teleport-API.patch [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
++     * Note: Teleporting to a different world with ignorePassengers to true while the entity has entities riding it
++     * will cause this teleportation to return false.
++     *
++     * @param location New location to teleport this entity to
++     * @param ignorePassengers If passengers should be ignored when teleporting
++     * @return <code>true</code> if the teleport was successful
++     */
++    default boolean teleport(@NotNull Location location, @NotNull TeleportCause cause, boolean ignorePassengers) {
++        return teleport(location, cause, ignorePassengers, true);
++    }
```
Comment (stefvanschie) [inline patches/api/0382-More-Teleport-API.patch]: The javadocs here is missing the teleport cause parameter.

File: patches/api/0382-More-Teleport-API.patch [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
++     * will cause this teleportation to return false.
++     * Note: Teleporting to a different world with dismount to false while this entity is riding another entity will
++     * cause this teleportation to return false.
++     *
++     * @param location New location to teleport this entity to
++     * @param ignorePassengers If passengers should be ignored when teleporting
++     * @param dismount If the entity should be dismounted if they are riding another entity
++     * @return <code>true</code> if the teleport was successful
++     */
++    boolean teleport(@NotNull Location location, @NotNull TeleportCause cause, boolean ignorePassengers, boolean dismount);
```
Comment (stefvanschie) [inline patches/api/0382-More-Teleport-API.patch]: Same as above, javadocs is missing the teleport cause parameter.

File: patches/api/0382-More-Teleport-API.patch [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
++     * Relative teleportation flags are only used client side, and cause the player to not lose velocity in that
++     * specific coordinate. The location of the teleportation will not change.
++     *
++     * @param location New location to teleport this entity to
++     * @param ignorePassengers If passengers should be ignored when teleporting
++     * @param dismount If the entity should be dismounted if they are riding another entity
++     * @param teleportFlags Coordinates of the location that the client should handle as relative teleportation
++     * @return <code>true</code> if the teleport was successful
++     */
++    boolean teleport(@NotNull Location location, @NotNull org.bukkit.event.player.PlayerTeleportEvent.TeleportCause cause, boolean ignorePassengers, boolean dismount, @NotNull io.papermc.paper.entity.RelativeTeleportFlag... teleportFlags);
```
Comment (stefvanschie) [inline patches/api/0382-More-Teleport-API.patch]: Same as above, teleport cause is missing from javadocs.

Maybe also introduce a NotNull annotation on the varargs itself: `@NotNull RelativeTeleportFlag @NotNull ... teleportFlags`?

File: patches/api/0382-More-Teleport-API.patch [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
++    }
++
++    /**
++     * Causes the player to look towards the given entity.
++     *
++     * @param entity Entity to look at
++     * @param playerAnchor What part of the player should face the entity
++     * @param entityAnchor What part of the entity the player should face
++     */
++    void lookAt(@NotNull org.bukkit.entity.Entity entity, @NotNull io.papermc.paper.entity.LookAnchor playerAnchor,  @NotNull io.papermc.paper.entity.LookAnchor entityAnchor);
```
Comment (stefvanschie) [inline patches/api/0382-More-Teleport-API.patch]: Very nitpicky, but there is a double space between the last two parameters.

File: patches/api/0382-More-Teleport-API.patch [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
++        teleportFlagSet = java.util.Collections.emptySet(); // Paper - Teleport API
+     }
+ 
+     public PlayerTeleportEvent(@NotNull final Player player, @NotNull final Location from, @Nullable final Location to, @NotNull final TeleportCause cause) {
+@@ -23,6 +29,16 @@ public class PlayerTeleportEvent extends PlayerMoveEvent {
+         this.cause = cause;
+     }
+ 
++    // Paper start - Teleport API
++    public PlayerTeleportEvent(@NotNull final Player player, @NotNull final Location from, @Nullable final Location to, @NotNull final TeleportCause cause, boolean dismounted, @NotNull java.util.Set<io.papermc.paper.entity.RelativeTeleportFlag> teleportFlagSet) {
```
Comment (stefvanschie) [inline patches/api/0382-More-Teleport-API.patch]: Maybe also specify the NotNull annotation on the generic type of the set here as well as in the method that returns these flags below?

**Review by stefvanschie — CHANGES_REQUESTED** [review]
You asked for feedback in Discord, from anyone who lurks, so... here it is. Some of the comments are about generics, not entirely certain what the consensus on those is, but worth considering I think? Regardless, this looks nice.

**Comment by lynxplay** [comment]
Functionality wise, the `setRotation` one works well for me in my tests with the removed multiplier :+1: I'll also try to go over the javadocs as well to join stefvanschie in javadoc analysis :)

File: patches/api/0382-More-Teleport-API.patch (hunk 18 above)
Comment (Owen1212055) [reply patches/api/0382-More-Teleport-API.patch]: image

I keep getting stuff like this, idk I haven't rly used it in this sense before.

**Comment by Owen1212055** [comment]
Thanks so much for the feedback, resolved :)

File: patches/api/0382-More-Teleport-API.patch (hunk 18 above)
Comment (stefvanschie) [reply patches/api/0382-More-Teleport-API.patch]: I believe that's because the full class name is used. From my testing, if you write `io.papermc.paper.entity.@NotNull RelativeTeleportFlag` inside the generic, it should work. Why you need to put the annotation right in the middle of the class name for generics, but can put it before the entire class name in other places, I have no idea.

**Review by electronicboy — APPROVED** [review]

**Comment by electronicboy** [comment]
generally looks fine, the only real nits I have are potentially JD related, describe the action, not the return; i.e. if the TP call can't be done due to state expectations, i.e. mounted & ignorePassengers, the teleport fails, not the method returns false, outside of that

**Comment by Owen1212055** [comment]
That’s how things should be, this pr used to do that however later on it was removed to boolean expressions only for backwards compatibility.

File: patches/api/0382-More-Teleport-API.patch [hunk 19]
```diff
@@ -0,0 +1,282 @@
...
++     * @return <code>true</code> if the teleport was successful
++     */
++    default boolean teleport(@NotNull Location location, @NotNull TeleportCause cause, boolean ignorePassengers) {
++        return teleport(location, cause, ignorePassengers, true);
++    }
++
++    /**
++     * Teleports this entity to the given location.
++     * <p>
++     * Note: Teleporting to a different world with ignorePassengers to true while the entity has entities riding it
```
Comment (electronicboy) [inline patches/api/0382-More-Teleport-API.patch]: I mean stuff like this, "return false" just sounds horrible and doesn't really describe that the teleportation will not occur

Comment (Owen1212055) [reply patches/api/0382-More-Teleport-API.patch]: Ahh I see, will do

**Comment by Alvinn8** [comment]
I have a plugin where I use nms for teleportation with passengers. I moved the code to use this pr instead and it's working great. Although at first, I didn't quite understand what the "ignorePassengers" parameter did and "If passengers should be ignored when teleporting" didn't really explain what ignored meant so I had to peek at the code. Maybe it should say something like whether to teleport the entity despite it having passengers. But otherwise this looks like great api 👍

**Comment by Owen1212055** [comment]
Rebased and documented ignore passengers, I'd appreciate for feedback @Alvinn8

File: patches/api/0382-More-Teleport-API.patch (hunk 19 above)
Comment (Owen1212055) [reply patches/api/0382-More-Teleport-API.patch]: I've added "and not occur". I'd appreciate feedback on this.

**Comment by 4drian3d** [comment]
Perhaps it would be good to replace the result of these methods with a CompletableFuture or to add corresponding teleportAsync methods that return a CompletableFuture, as well as the methods implemented by Paper in the same Entity interface https://papermc.io/javadocs/paper/1.18/org/bukkit/entity/Entity.html#teleportAsync(org.bukkit.Location,org.bukkit.event.player.PlayerTeleportEvent.TeleportCause)

**Comment by Owen1212055** [comment]
I wasn't sure because it's just really simple to do yourself if you ever need to get this specific.

File: patches/api/0384-More-Teleport-API.patch [hunk 20]
```diff
@@ -0,0 +1,282 @@
...
+ 
++    // Paper start - Teleport API
++    /**
++     * Teleports this entity to the given location.
++     * <p>
++     * Note: Teleporting to a different world with ignorePassengers to true while the entity has entities riding it
++     * will cause this teleportation to return false and not occur.
++     *
++     * @param location New location to teleport this entity to
++     * @param ignorePassengers If all passengers should be required to be removed prior to teleportation
```
Comment (Alvinn8) [inline patches/api/0384-More-Teleport-API.patch]: I interpret this as reversed now. I get the impression that setting ignorePassengers to true would require removing the passengers, but it's the other way around.

**Comment by Owen1212055** [comment]
Rebase

**Comment by Owen1212055** [comment]
Rebased, all methods and classes have been marked with @org.jetbrains.annotations.ApiStatus.Experimental

**Comment by Owen1212055** [comment]
Rebased....

**Comment by Owen1212055** [comment]
Rebased!
